File: apps/android/app/src/main/java/app/bananatape/mobile/adapters/FakeAdapters.kt
package app.bananatape.mobile.adapters

import java.nio.file.Files
import java.nio.file.Path
import java.time.Duration

class FakeProjectStorage(
    projects: List<MobileProjectRecord> = emptyList(),
    private val corruptProjectIds: Set<String> = emptySet(),
) : ProjectStorage {
    private val projectsById: MutableMap<String, MobileProjectRecord> = projects.associateBy { it.id }.toMutableMap()

    override fun create(project: MobileProjectRecord): AdapterResult<MobileProjectRecord> {
        projectsById[project.id] = project
        return AdapterResult.Success(project)
    }

    override fun read(id: String): AdapterResult<MobileProjectRecord> {
        if (corruptProjectIds.contains(id)) {
            return AdapterResult.Failure(AdapterError.CorruptProject(id))
        }
        val project = projectsById[id] ?: return AdapterResult.Failure(AdapterError.StorageNotFound(id))
        return AdapterResult.Success(project)
    }

    override fun list(): List<MobileProjectSummary> = projectsById.values
        .map { MobileProjectSummary(id = it.id, name = it.name) }
        .sortedBy { it.name }

    override fun delete(id: String): AdapterResult<Unit> {
        projectsById.remove(id) ?: return AdapterResult.Failure(AdapterError.StorageNotFound(id))
        return AdapterResult.Success(Unit)
    }

    override fun rename(id: String, name: String): AdapterResult<MobileProjectRecord> {
        val project = projectsById[id] ?: return AdapterResult.Failure(AdapterError.StorageNotFound(id))
        val updated = project.copy(name = name)
        projectsById[id] = updated
        return AdapterResult.Success(updated)
    }
}

class FakePermissionGateway(
    private val decisions: Map<PermissionScope, PermissionDecision>,
) : PermissionGateway {
    override fun decision(scope: PermissionScope): PermissionDecision = decisions[scope] ?: PermissionDecision.GRANTED
}

class FakeNetworkStatus(private val reachability: NetworkReachability) : NetworkStatus {
    override fun currentReachability(): NetworkReachability = reachability
}

class FakeProviderAuth(
    private val availabilityByProvider: Map<ProviderId, ProviderAvailability>,
    private val networkStatus: NetworkStatus,
) : ProviderAuth {
    override fun availability(provider: ProviderId): ProviderAvailability {
        if (provider == ProviderId.CODEX) {
            return ProviderAvailability.UNAVAILABLE
        }
        if (networkStatus.currentReachability() == NetworkReachability.OFFLINE) {
            return ProviderAvailability.OFFLINE
        }
        return availabilityByProvider[provider] ?: ProviderAvailability.MISSING_KEY
    }
}

class FakeImageMemoryPolicy(private val maxBytes: Int) : ImageMemoryPolicy {
    override fun evaluate(byteCount: Int): AdapterResult<Unit> {
        if (byteCount > maxBytes) {
            return AdapterResult.Failure(AdapterError.OversizedImage(maxBytes = maxBytes, actualBytes = byteCount))
        }
        return AdapterResult.Success(Unit)
    }
}

class FakeImageImport(
    private val permissionGateway: PermissionGateway,
    private val memoryPolicy: ImageMemoryPolicy,
) : ImageImport {
    override fun importImage(id: String, mimeType: ImageMimeType, byteCount: Int): AdapterResult<ImportedImage> {
        if (permissionGateway.decision(PermissionScope.IMAGE_IMPORT) == PermissionDecision.DENIED) {
            return AdapterResult.Failure(AdapterError.PermissionDenied(PermissionScope.IMAGE_IMPORT))
        }
        return when (mimeType) {
            ImageMimeType.PNG,
            ImageMimeType.JPEG,
            -> when (val memoryResult = memoryPolicy.evaluate(byteCount)) {
                is AdapterResult.Success -> AdapterResult.Success(ImportedImage(id = id, mimeType = mimeType, byteCount = byteCount))
                is AdapterResult.Failure -> memoryResult
            }
            ImageMimeType.WEBP,
            ImageMimeType.GIF,
            ImageMimeType.HEIC,
            -> AdapterResult.Failure(AdapterError.UnsupportedFileType(mimeType))
        }
    }
}

class FakeImageExport(private val permissionGateway: PermissionGateway) : ImageExport {
    override fun exportImage(id: String, destination: ImageExportDestination, byteCount: Int): AdapterResult<ExportedImage> {
        if (permissionGateway.decision(PermissionScope.IMAGE_EXPORT) == PermissionDecision.DENIED) {
            return AdapterResult.Failure(AdapterError.PermissionDenied(PermissionScope.IMAGE_EXPORT))
        }
        return AdapterResult.Success(ExportedImage(id = id, destination = destination, byteCount = byteCount))
    }
}

class FakeGalleryImageExport(
    private val permissionGateway: PermissionGateway,
    private val relativePath: String = "Pictures/BananaTape/",
) : GalleryImageExport {
    override fun saveToGallery(image: ExportableImage): AdapterResult<GalleryExportReceipt> {
        if (permissionGateway.decision(PermissionScope.IMAGE_EXPORT) == PermissionDecision.DENIED) {
            return AdapterResult.Failure(AdapterError.PermissionDenied(PermissionScope.IMAGE_EXPORT))
        }
        return AdapterResult.Success(
            GalleryExportReceipt(
                id = image.id,
                relativePath = relativePath,
                displayName = displayName(image),
                mimeType = image.mimeType,
                width = image.width,
                height = image.height,
                byteCount = image.byteCount,
                createdAt = image.createdAt,
            ),
        )
    }

    private fun displayName(image: ExportableImage): String = "${image.id}${extension(image.mimeType)}"

    private fun extension(mimeType: ImageMimeType): String = when (mimeType) {
        ImageMimeType.PNG -> ".png"
        ImageMimeType.JPEG -> ".jpg"
        ImageMimeType.WEBP,
        ImageMimeType.GIF,
        ImageMimeType.HEIC,
        -> ".img"
    }
}

class FakeOutboundImageShare(
    private val tempDirectory: Path,
    private val ttl: Duration,
) : OutboundImageShare {
    override fun prepareShare(image: ExportableImage): AdapterResult<SharedImage> = runCatching {
        Files.createDirectories(tempDirectory)
        val sharePath = tempDirectory.resolve("${image.id}${extension(image.mimeType)}")
        Files.copy(image.filePath, sharePath, java.nio.file.StandardCopyOption.REPLACE_EXISTING)
        SharedImage(
            id = image.id,
            contentUri = "content://app.bananatape.mobile.share/${sharePath.fileName}",
            mimeType = image.mimeType,
            byteCount = image.byteCount,
            expiresAt = image.createdAt.plus(ttl),
        )
    }.fold(
        onSuccess = { AdapterResult.Success(it) },
        onFailure = { AdapterResult.Failure(AdapterError.CorruptProject(image.id)) },
    )

    private fun extension(mimeType: ImageMimeType): String = when (mimeType) {
        ImageMimeType.PNG -> ".png"
        ImageMimeType.JPEG -> ".jpg"
        ImageMimeType.WEBP,
        ImageMimeType.GIF,
        ImageMimeType.HEIC,
        -> ".img"
    }
}

class InMemoryEncryptedKeyValueStore(
    initialValues: Map<String, String> = emptyMap(),
) : EncryptedKeyValueStore {
    private val values = initialValues.toMutableMap()

    override fun readString(key: String): String? = values[key]

    override fun writeString(key: String, value: String) {
        values[key] = value
    }

    override fun remove(key: String) {
        values.remove(key)
    }
}

class InMemoryOpenAiApiKeyStore(key: String? = null) : OpenAiApiKeyStore {
    private val encryptedStore = InMemoryEncryptedKeyValueStore(
        key?.let { mapOf(EncryptedPreferencesOpenAiApiKeyStore.OpenAiKeyName to it) } ?: emptyMap(),
    )
    private val delegate = EncryptedPreferencesOpenAiApiKeyStore(encryptedStore)

    override fun apiKeyState(): OpenAiApiKeyState = delegate.apiKeyState()
    override fun readApiKey(): String? = delegate.readApiKey()
    override fun saveApiKey(key: String): AdapterResult<Unit> = delegate.saveApiKey(key)
    override fun deleteApiKey(): AdapterResult<Unit> = delegate.deleteApiKey()
}

File: apps/ios/BananaTape/BananaTape/FakeAdapters.swift
import Foundation

final class FakeProjectStorage: ProjectStorage {
    private var projects: [String: MobileProjectRecord]
    private let corruptProjectIDs: Set<String>

    init(projects: [MobileProjectRecord] = [], corruptProjectIDs: Set<String> = []) {
        self.projects = Dictionary(uniqueKeysWithValues: projects.map { ($0.id, $0) })
        self.corruptProjectIDs = corruptProjectIDs
    }

    func create(_ project: MobileProjectRecord) -> Result<MobileProjectRecord, AdapterError> {
        projects[project.id] = project
        return .success(project)
    }

    func read(id: String) -> Result<MobileProjectRecord, AdapterError> {
        if corruptProjectIDs.contains(id) {
            return .failure(.corruptProject(id))
        }
        guard let project = projects[id] else {
            return .failure(.storageNotFound(id))
        }
        return .success(project)
    }

    func list() -> [MobileProjectSummary] {
        projects.values
            .map { MobileProjectSummary(id: $0.id, name: $0.name) }
            .sorted { $0.name < $1.name }
    }

    func delete(id: String) -> Result<Void, AdapterError> {
        guard projects.removeValue(forKey: id) != nil else {
            return .failure(.storageNotFound(id))
        }
        return .success(())
    }

    func renameProject(id: String, name: String) -> Result<MobileProjectRecord, AdapterError> {
        guard let project = projects[id] else { return .failure(.storageNotFound(id)) }
        let updated = MobileProjectRecord(id: id, name: name, manifestJSON: project.manifestJSON.replacingOccurrences(of: "\"name\": \"\(project.name)\"", with: "\"name\": \"\(name)\""), historyJSON: project.historyJSON, canvasJSON: project.canvasJSON)
        projects[id] = updated
        return .success(updated)
    }
}

struct FakePermissionGateway: PermissionGateway {
    let decisions: [PermissionScope: PermissionDecision]

    func decision(for scope: PermissionScope) -> PermissionDecision {
        decisions[scope, default: .granted]
    }
}

struct FakeNetworkStatus: NetworkStatus {
    let reachability: NetworkReachability

    func currentReachability() -> NetworkReachability {
        reachability
    }
}

struct FakeProviderAuth: ProviderAuth {
    let availabilityByProvider: [ProviderID: ProviderAvailability]
    let networkStatus: NetworkStatus

    func availability(for provider: ProviderID) -> ProviderAvailability {
        if provider == .codex {
            return .unavailable
        }
        if networkStatus.currentReachability() == .offline {
            return .offline
        }
        return availabilityByProvider[provider, default: .missingKey]
    }
}

struct FakeImageMemoryPolicy: ImageMemoryPolicy {
    let maxBytes: Int

    func evaluate(byteCount: Int) -> Result<Void, AdapterError> {
        if byteCount > maxBytes {
            return .failure(.oversizedImage(maxBytes: maxBytes, actualBytes: byteCount))
        }
        return .success(())
    }
}

struct FakeImageImport: ImageImport {
    let permissionGateway: PermissionGateway
    let memoryPolicy: ImageMemoryPolicy

    func importImage(id: String, mimeType: ImageMimeType, byteCount: Int) -> Result<ImportedImage, AdapterError> {
        if permissionGateway.decision(for: .imageImport) == .denied {
            return .failure(.permissionDenied(.imageImport))
        }
        switch mimeType {
        case .png, .jpeg:
            return memoryPolicy.evaluate(byteCount: byteCount).map { ImportedImage(id: id, mimeType: mimeType, byteCount: byteCount) }
        case .webp, .gif, .heic:
            return .failure(.unsupportedFileType(mimeType))
        }
    }
}

struct FakeImageExport: ImageExport {
    let permissionGateway: PermissionGateway

    func exportImage(id: String, destination: ImageExportDestination, byteCount: Int) -> Result<ExportedImage, AdapterError> {
        if permissionGateway.decision(for: .imageExport) == .denied {
            return .failure(.permissionDenied(.imageExport))
        }
        return .success(ExportedImage(id: id, destination: destination, byteCount: byteCount))
    }
}

final class InMemoryOpenAIAPIKeyStore: OpenAIAPIKeyStore {
    private var key: String?

    init(key: String? = nil) {
        self.key = key
    }

    func apiKeyState() -> OpenAIAPIKeyState {
        guard let key, !key.isEmpty else { return .missing }
        return .present(maskedSuffix: String(key.suffix(4)))
    }

    func readAPIKey() -> String? { key }

    func saveAPIKey(_ key: String) -> Result<Void, AdapterError> {
        self.key = key
        return .success(())
    }

    func deleteAPIKey() -> Result<Void, AdapterError> {
        key = nil
        return .success(())
    }
}
